lib/iaas/GcpClient.js

```diff
@@ -0,0 +1,200 @@
+'use strict';
+const _ = require('lodash');
+const Promise = require('bluebird');
+const GcpStorage = require('@google-cloud/storage');
+const logger = require('../logger');
+const errors = require('../errors');
+const utils = require('../utils');
+const ComputeClient = require('./ComputeClient');
+const BaseCloudClient = require('./BaseCloudClient');
+const NotFound = errors.NotFound;
+const Unauthorized = errors.Unauthorized;
+const Forbidden = errors.Forbidden;
+
+class GcpClient extends BaseCloudClient {
+  constructor(settings) {
+    super(settings);
+    this.constructor.validateParams(_.chain(this.settings).value());
+    this.storageClient = this.constructor.createStorageClient(_
+      .chain(this.settings)
+      .omit('name')
+      .set('provider', this.provider)
+      .value()
+    );
+    this.computeClient = this.constructor.createComputeClient(_
+      .chain(this.settings)
+      .set('provider', this.provider)
+      .value()
+    );
+  }
+
+  getContainer(container) {
+    if (arguments.length < 1) {
+      container = this.containerName;
+    }
+
+    return this.storageClient
+      .bucket(container)
+      .get();
+  }
+
+  list(container, options) {
+    if (arguments.length < 2) {
+      options = container;
+      container = this.containerName;
+    }
+    const prefix = options ? options.prefix : null;
+    const queryOptions = {};
+    if (prefix) {
+      queryOptions.prefix = prefix;
+    }
+    queryOptions.autoPaginate = (options && options.autoPaginate) ? options.autoPaginate : true;
+
+    return Promise.try(() => {
+      return this.storageClient
+        .bucket(container)
+        .getFiles(queryOptions)
+        .then(results => {
+          const resultFiles = results[0];
+          const files = []; 
+          _.each(resultFiles, file => files.push(_       
+            .chain(file)       
+            .pick('name')       
+            .value()      
+          ));   
+          return files;
+        });
+    });
+  }
+
+  remove(container, file) {
+    if (arguments.length < 2) {
+      file = container;
+      container = this.containerName;
+    }
+    logger.debug(`Deleting file ${file} in container ${container} `);
+    return Promise.try(() => {
+        return this.storageClient
+          .bucket(container)
+          .file(file)
+          .delete();
+      })
+      .catch(BaseCloudClient.providerErrorTypes.Unauthorized, err => {
+        logger.error(err.message);
```

Comment:
when using logger.error, always provide the error object so that stack trace gets printed. So you could change the above to 
```
logger.error(`Authorization at google cloud storage provider failed while deleting blob ${file} in container ${container}`, err);
```

Reply:
Since I am throwing error back, will not the stack trace be printed in parent classes? If I also log it here, debugging through logs will be difficult as same errors/stack traces will get printed multiple times at multiple places.